File: 82-remove-duplicates-from-sorted-list-ii/remove-duplicates-from-sorted-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        noDuplicates = []

        curr = head
        prev = None
        while curr:
            hasDuplicate = False
            nextNode = curr.next
            if prev and prev.val == curr.val:
                hasDuplicate = True
            if nextNode and nextNode.val == curr.val:
                hasDuplicate = True
            
            if not hasDuplicate:
                noDuplicates.append(curr)
            prev = curr
            curr = curr.next

        for i in range(len(noDuplicates) - 1):
            curr = noDuplicates[i]
            curr.next = noDuplicates[i + 1]

        if len(noDuplicates) == 0: return None

        lastNode = noDuplicates[-1]
        if lastNode:
            lastNode.next = None
        return noDuplicates[0]

File: 82-remove-duplicates-from-sorted-list-ii/remove-duplicates-from-sorted-list-ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteDuplicates = function(head) {
    const noDuplicates = [];

    let curr = head;
    let prev = null;
    while (curr) {
        let hasDuplicate = false;
        const nextNode = curr.next;
        
        if (prev && prev.val === curr.val) hasDuplicate = true;
        if (nextNode && nextNode.val === curr.val) hasDuplicate = true;
        
        if (!hasDuplicate) noDuplicates.push(curr)
        prev = curr;
        curr = curr.next
    }

    if (noDuplicates.length === 0) return null;

    for (let i = 0; i < noDuplicates.length - 1; i++) {
        const curr = noDuplicates[i];
        curr.next = noDuplicates[i + 1];
    }
    const lastNode = noDuplicates[noDuplicates.length - 1];
    lastNode.next = null;
    return noDuplicates[0]
};